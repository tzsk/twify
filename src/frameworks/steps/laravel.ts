import chalk from 'chalk';
import fs from 'fs-extra';
import path from 'path';

export async function setupWelcomePage() {
  const filename = 'welcome.blade.php';
  const welcomeFile = path.join(process.cwd(), `resources/views`, filename);
  const welcomePage = await fs.readFile(welcomeFile, 'utf8');

  console.log(`\n- Setting up ${chalk.blue.bold(filename)}...`);

  await fs.writeFile(
    welcomeFile,
    welcomePage.replace(
      /<\/head>/,
      "\t@vite(['resources/css/app.css'])\n\t</head>"
    )
  );
}
