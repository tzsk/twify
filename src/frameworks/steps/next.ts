import glob from 'glob';
import util from 'util';

export async function resolveCssLocation() {
  const [match] = await util.promisify(glob)('./**/globals.css');

  if (!match) {
    return './styles/globals.css';
  }

  return match;
}
