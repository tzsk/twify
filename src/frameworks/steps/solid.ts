import glob from 'glob';
import util from 'util';

export async function resolveCssLocation() {
  const [match] = await util.promisify(glob)('./src/**/root.css', {
    ignore: ['node_modules'],
  });

  if (!match) {
    return './src/index.css';
  }

  return match;
}
