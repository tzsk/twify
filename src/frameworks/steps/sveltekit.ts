import fs from 'fs-extra';
import path from 'path';
import j from 'jscodeshift';
import chalk from 'chalk';

export async function movePostCSS() {
  const source = path.join(process.cwd(), 'postcss.config.js');
  const destination = path.join(process.cwd(), 'postcss.config.cjs');
  if (fs.existsSync(source)) {
    await fs.move(source, destination, { overwrite: true });
  }
}

export async function setupLayoutFile() {
  const ts = fs.existsSync(path.join(process.cwd(), 'tsconfig.json'));
  const location = 'src/routes/+layout.svelte';
  const file = path.join(process.cwd(), location);

  if (fs.existsSync(file)) {
    console.log(`\n${chalk.blue.bold(location)} already exists.`);
    return;
  }

  console.log(`\n- Creating ${chalk.blue.bold(location)}...`);
  await fs.ensureFile(file);

  const opening = ts ? `<script lang="ts">` : `<script>`;
  const code = `${opening}
  import "../app.css";
</script>

<slot />
`;

  await fs.writeFile(file, code);
}
