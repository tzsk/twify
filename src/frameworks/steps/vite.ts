import chalk from 'chalk';
import fs from 'fs-extra';
import path from 'path';
import j from 'jscodeshift';

function _withCssImport(code: string): string {
  const ast = j.withParser('flow')(code);

  const imports = ast.find(j.ImportDeclaration);

  imports
    .at(imports.length - 1)
    .insertAfter(j.importDeclaration([], j.literal('./style.css')));

  return ast.toSource();
}

export async function setupMainFile() {
  const ts = fs.existsSync(path.join(process.cwd(), 'tsconfig.json'));
  let filename: string = ts ? 'main.ts' : 'main.js';
  filename = fs.existsSync(path.join(process.cwd(), 'src', filename))
    ? filename
    : `${filename}x`;

  console.log(`\n- Setting up ${chalk.blue.bold(`src/${filename}`)}...`);
  const filePath = path.join(process.cwd(), 'src', filename);

  const code = await fs.readFile(filePath, 'utf8');

  await fs.writeFile(filePath, _withCssImport(code));
}
