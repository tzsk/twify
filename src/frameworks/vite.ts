import { Framework } from '../types';
import { setupMainFile } from './steps/vite';

const Vite: Framework = {
  requiredDependencies: ['tailwindcss', 'postcss', 'autoprefixer'],
  initCommands: ['npx tailwindcss init -p'],
  cssLocation: './src/style.css',
  content: {
    name: 'tailwind.config.js',
    files: ['./index.html', './src/**/*.{vue,js,ts,jsx,tsx}'],
  },
  steps: [setupMainFile],
};

export default Vite;
