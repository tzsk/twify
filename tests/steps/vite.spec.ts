import fs from 'fs-extra';
import { setupMainFile } from '../../src/frameworks/steps/vite';

async function runStep(file: string) {
  const code = fs.readFileSync(`./tests/fixtures/vite/${file}`, 'utf8');
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  vi.spyOn(fs, 'readFile').mockResolvedValue(code as any);
  const writeSpy = vi.spyOn(fs, 'writeFile').mockResolvedValue();

  await setupMainFile();

  const [fileName, content] = writeSpy.mock.lastCall || [];

  return { fileName, content };
}

describe('Vite Steps', () => {
  beforeAll(() => {
    vi.stubGlobal('console', { ...console, log: vi.fn() });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  afterAll(() => {
    vi.resetAllMocks();
  });

  it('can setup vite vue js project', async () => {
    vi.spyOn(fs, 'existsSync')
      .mockReturnValueOnce(false)
      .mockReturnValueOnce(true);
    const { fileName, content } = await runStep('main.vue.stub');

    expect(fileName).toMatch('main.js');
    expect(content).toMatchSnapshot();
  });

  it('can setup vite vue ts project', async () => {
    vi.spyOn(fs, 'existsSync')
      .mockReturnValueOnce(true)
      .mockReturnValueOnce(true);
    const { fileName, content } = await runStep('main.vue.stub');

    expect(fileName).toMatch('main.ts');
    expect(content).toMatchSnapshot();
  });

  it('can setup vite jsx project', async () => {
    vi.spyOn(fs, 'existsSync')
      .mockReturnValueOnce(false)
      .mockReturnValueOnce(false);
    const { fileName, content } = await runStep('main.react.stubx');

    expect(fileName).toMatch('main.jsx');
    expect(content).toMatchSnapshot();
  });

  it('can setup vite tsx project', async () => {
    vi.spyOn(fs, 'existsSync')
      .mockReturnValueOnce(true)
      .mockReturnValueOnce(false);
    const { fileName, content } = await runStep('main.react.stubx');

    expect(fileName).toMatch('main.tsx');
    expect(content).toMatchSnapshot();
  });
});
