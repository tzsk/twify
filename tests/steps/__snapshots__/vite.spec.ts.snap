// Vitest Snapshot v1

exports[`Vite Steps > can setup vite jsx project 1`] = `
"import React from 'react'
import ReactDOM from 'react-dom/client'
import App from './App'
import './index.css'

import \\"./style.css\\";

ReactDOM.createRoot(document.getElementById('root')).render(
  <React.StrictMode>
    <App />
  </React.StrictMode>
)"
`;

exports[`Vite Steps > can setup vite tsx project 1`] = `
"import React from 'react'
import ReactDOM from 'react-dom/client'
import App from './App'
import './index.css'

import \\"./style.css\\";

ReactDOM.createRoot(document.getElementById('root')).render(
  <React.StrictMode>
    <App />
  </React.StrictMode>
)"
`;

exports[`Vite Steps > can setup vite vue js project 1`] = `
"import { createApp } from 'vue'
import App from './App.vue'

import \\"./style.css\\";

createApp(App).mount('#app')
"
`;

exports[`Vite Steps > can setup vite vue ts project 1`] = `
"import { createApp } from 'vue'
import App from './App.vue'

import \\"./style.css\\";

createApp(App).mount('#app')
"
`;
